lewis/adapters/modbus.py

```diff
@@ -0,0 +1,569 @@
+        :return: ModbusTCPFrame response to the request
+        """
+        addr, bit_count, byte_count = struct.unpack('>HHB', bytes(request.data[:5]))
+        data = request.data[5:]
+
+        if not 0x0001 <= bit_count <= 0x07B0 or byte_count != ceil(bit_count / 8):
+            return request.create_exception(MBEX.DATA_VALUE)
+
+        # Bytes to bits: first byte -> last byte, LSB -> MSB
+        bits = [False] * bit_count
+        for i in range(bit_count):
+            bits[i] = bool(data[i // 8] & (1 << i % 8))
+
+        try:
+            self._datastore.co.set(addr, bits)
+        except IndexError:
+            return request.create_exception(MBEX.DATA_ADDRESS)
+
+        # Respond to confirm
+        return request.create_response(request.data[:4])
+
+    def _handle_write_multiple_registers(self, request):
+        """
+        Handle request as per Modbus Application Protocol v1.1b3:
+        Section 6.12 - (0x10) Write Multiple registers
+
+        :param request: ModbusTCPFrame containing the request
+        :return: ModbusTCPFrame response to the request
+        """
+        addr, reg_count, byte_count = struct.unpack('>HHB', bytes(request.data[:5]))
+        data = request.data[5:]
+
+        if not 0x0001 <= reg_count <= 0x007B or byte_count != reg_count * 2:
+            return request.create_exception(MBEX.DATA_VALUE)
+
+        try:
+            words = list(struct.unpack('>%dH' % reg_count, data))
+            self._datastore.hr.set(addr, words)
+        except IndexError:
+            return request.create_exception(MBEX.DATA_ADDRESS)
+
+        # Respond to confirm
+        return request.create_response(request.data[:4])
+
+
+class ModbusHandler(asyncore.dispatcher_with_send):
+    def __init__(self, sock, adapter, server):
+        asyncore.dispatcher_with_send.__init__(self, sock=sock)
+        self._datastore = ModbusDataStore(adapter.di, adapter.co, adapter.ir, adapter.hr)
+        self._modbus = ModbusProtocol(self.send, self._datastore)
+        self._server = server
+
+    def handle_read(self):
+        data = self.recv(8192)
+        self._modbus.process(data)
+
+    def handle_close(self):
+        self._server.remove_handler(self)
+        self.close()
+
+
+class ModbusServer(asyncore.dispatcher):
+    def __init__(self, host, port, adapter=None):
+        asyncore.dispatcher.__init__(self)
+        self.adapter = adapter
+        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.set_reuse_addr()
+        self.bind((host, port))
+        self.listen(5)
+        self._accepted_connections = []
+
+    def handle_accept(self):
+        pair = self.accept()
+        if pair is not None:
+            sock, _ = pair
+            handler = ModbusHandler(sock, self.adapter, self)
+            self._accepted_connections.append(handler)
+
+    def remove_handler(self, handler):
+        self._accepted_connections.remove(handler)
+
+    def handle_close(self):
```

Comment:
Yeah, I'm not sure which way is more correct... or whether there's any difference at all. Agreed, something to look at when we refactor here. Best to leave as is to serve as a reminder that this is something to look into.